### ISSUE TITLE
Periodic/Nonperiodic keywords to selections

### ISSUE THREAD

**Issue by richardjgowers** [MDAnalysis/mdanalysis#759]
So currently we have some [geometric selections](http://pythonhosted.org/MDAnalysis/documentation_pages/selections.html#geometric) which allow you to select atoms based on absolute and relative positions.  These can work with and without periodic boundary conditions, both are useful at different times.  

The use of PBC is currently controlled by flags, which I think we're moving towards removing (#315) so instead a keyword in the selection could be used.

```
u.select_atoms('A around B distance C selection D')

u.select_atoms('A point B x y z C distance D')
```

So with the above selections, there are positions **A** -> **D** where we could put the keyword.  

Position **A** would be tricky as the keyword precedes the token, but it could be done.  (`'periodic around 7.0 protein'`)

**D** is messy for the `around` case as generally anything after **C** gets made into another selection.

I think **C** reads best.. (`'around 7.0 periodic protein'`)

So tl;dr
- add `periodic` keyword?
- what should the default be? (probably same as currently?)
- where to put the `periodic` keyword?

**Comment by kain88-de** [comment]
@richardjgowers  I'm sorry I think something when wrong with your formating. Are **A** .. **D** supposed to be options to choose from or atoms?

**Comment by kain88-de** [comment]
I might replace `periodic` with `pbc`

**Comment by richardjgowers** [comment]
They're places in the selection string we could insert the keyword, so options

**Comment by jbarnoud** [comment]
I do not have a strong opinion on the question, but I'll go with **D**. It makes the keyword easier to omit, and existing code will not look odd when compared to new one.

@kain88-de Why would `pbc` be better than `periodic`?

What should be the opposite keyword? `noperiodic` (or `nopbc`)?

**Comment by richardjgowers** [comment]
@jbarnoud the problem with **D** is that the periodic keyword can end up a long way away (as selection can be arbitrarily long).. so
```
'around 7.0 protein periodic'  # this reads well!

'around 7.0 not resid 7 or name OH periodic'  # this is getting worse?
```

And then the implementation problem is usually those selections read the distance (7.0 here) [then pass all following tokens to be parsed again as the selection.](https://github.com/MDAnalysis/mdanalysis/blob/05a8aa600915290596a70f983350bcc45517fc41/package/MDAnalysis/core/Selection.py#L249)  So we'd need some sort of way of passing back of the periodic keyword, not impossible though.

nonperiodic / nopbc is the opposite

**Comment by jbarnoud** [comment]
@richardjgowers I see your point. I do not like **A** because I much prefer the selection keyword to be first (much easier to parse, even for a human). I think the `periodic` keyword applies to the distance (or the coordinate) so **C** would make sense.

**Comment by mnmelo** [comment]
I for one think **D** is ok. It might also be better to have it as a separate `kwarg` instead:
```python
u.select_atoms('A around B distance C selection', pbc=True)
```

One other aspect is that one might want different PBC treatment for different subselections in the expression. **C** would be best for that, as we can have one per subselection.

**Comment by richardjgowers** [comment]
Ahhh, I didn't even think of a kwarg, maybe that's even better.  I'm assuming that it doesn't mess with the current usage unless you have a group called pbc?

**Comment by mnmelo** [comment]
Exactly. I think it's ok to have a couple of reserved words for this and other purposes (see @jbarnoud's [comment](https://github.com/MDAnalysis/mdanalysis/issues/175#issuecomment-184390345) in #175).

**Comment by jbarnoud** [comment]
@mnmelo Are we sure there is no usecase where we would like to mix periodic and non periodic selection keywords?

What about `around 7.0 periodic of protein and not around 10.0 nonperiodic of resname POPC`? Why would we want to prevent that?

**Comment by mnmelo** [comment]
Yea, if we want to keep that flexibility, then **C** is definitely the way to go.

**Comment by orbeckst** [comment]
I like the kwarg solution best for clarity and I think it covers the overwhelming majority of use cases; in fact, you could argue that you are more likely to get "bad" selections by inadvertently mixing periodic and non-periodic selections. It does occur a flexibility penalty but I can't think of any use cases where you'd need to mix. I am pretty sure you could get a mixed selection by using two selections and groups.

I also favor the *periodic* keyword (more explicit than *pbc*).

**Comment by richardjgowers** [comment]
Ok so I think this boils down to, can we think of a use case for wanting half periodic and half non periodic selections.  And then, if there is a case, is it so obscure that we can just say use two calls to select_atoms?

**Comment by jbarnoud** [comment]
On 19/03/16 10:44, Richard Gowers wrote:
>
> Ok so I think this boils down to, can we think of a use case for 
> wanting half periodic and half non periodic selections. And then, if 
> there is a case, is it so obscure that we can just say use two calls 
> to select_atoms?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/issues/759#issuecomment-198675236>
>
I though about it, but I could not find any.

**Comment by kain88-de** [comment]
I may have an application that needs a mix in the next 2-3 month but my guess is that we there can live with a `periodic` keyword comfortably as well.

**Comment by orbeckst** [comment]
I'd like to propose a separate keyword `pbc=True` for `select_atoms()` to enable PBC-aware distance calculations in selections:
```python
with_pbc = u.select_atoms('byres (resname SOL and around 4.0 protein)', pbc=True)
no_pbc = u.select_atoms('byres (resname SOL and around 4.0 protein)', pbc=False)
```
* I chose `pbc` because we are already using it for various methods of `AtomGroup`.
* The keyword applies to *all* selections that are performed by `select_atoms()`, including use of the OR selection when using multiple selection strings.
* The default should be `pbc=True` because if a PBC box is available, this choice is more likely to be correct (although it will incur a performance penalty and we can debate this choice).
* If no `Timestep.dimensions` is available, always switch to `pbc=False` (and log a warning if `pbc=True` was set).

**Comment by orbeckst** [comment]
In order to make progress with #782 we need to decide on how to indicate periodicity in selections. 

I'd like to go with my proposal https://github.com/MDAnalysis/mdanalysis/issues/759#issuecomment-199129343 with a kwarg for `select_atoms(..., pbc=True)` that enables or disables PBC for the complete selection. 

Aye/Nay ?

**Comment by kain88-de** [comment]
Aye. But I’m for keeping the current default of `pbc=False`.

**Comment by orbeckst** [comment]
Ok, I am fine with keeping the existing default (which is probably faster, too). I am amending the comment above.

**Comment by kain88-de** [comment]
> Ok, I am fine with keeping the existing default (which is probably faster, too)

It will also break less user code unnecessarily.
